src/simdutf/flb_simdutf_connector.cpp

```diff
@@ -175,45 +176,59 @@ int flb_simdutf_connector_convert_from_unicode(int preferred_encoding,
         return FLB_SIMDUTF_CONNECTOR_CONVERT_NOP;
     }
     else if ((encoding & simdutf::encoding_type::UTF16_LE) == simdutf::encoding_type::UTF16_LE) {
+        /* Skip the UTF-16 BOM */
+        if (length >= 2 && (input[0] & 0xFF) == 0xFF && (input[1] & 0xFF) == 0xFE) {
+            input += 2;
+            length -= 2;
+        }
+
         len = length;
         if (len % 2) {
             len--;
         }
         if (len < 2) {
             return FLB_SIMDUTF_CONNECTOR_CONVERT_NOP;
         }
-        for (i = 0 ; i < len;) {
-            if (i + 2 > len) {
-                break;
-            }
-            /* little-endian */
-            int lo = input[i++] & 0xFF;
-            int hi = input[i++] & 0xFF;
-            str16.push_back(hi << 8 | lo);
+
+        /* Check alignment to determine whether to copy or not */
+        if ((uintptr_t)input % 2 == 0) {
+            aligned_input = (const char16_t *)input;
+        }
+        else {
+            str16.resize(len / 2);
+            memcpy(str16.data(), input, len);
+            aligned_input = str16.data();
         }
 
-        return flb_simdutf_connector_convert_utf16le_to_utf8(str16.c_str(), str16.size(),
+        return flb_simdutf_connector_convert_utf16le_to_utf8(aligned_input, len / 2,
                                                              output, out_size);
     }
     else if ((encoding & simdutf::encoding_type::UTF16_BE) == simdutf::encoding_type::UTF16_BE) {
+        /* Skip the UTF-16 BOM */
+        if (length >= 2 && (input[0] & 0xFF) == 0xFE && (input[1] & 0xFF) == 0xFF) {
+            input += 2;
+            length -= 2;
+        }
+
         len = length;
         if (len % 2) {
             len--;
         }
         if (len < 2) {
             return FLB_SIMDUTF_CONNECTOR_CONVERT_NOP;
         }
-        for (i = 0; i < len;) {
-            if (i + 2 > len) {
-                break;
-            }
-            /* big-endian */
-            int lo = input[i++] & 0xFF;
-            int hi = input[i++] & 0xFF;
-            str16.push_back(lo | hi << 8);
+
+        /* Check alignment to determine whether to copy or not */
+        if ((uintptr_t)input % 2 == 0) {
+            aligned_input = (const char16_t *)input;
+        }
+        else {
+            str16.resize(len / 2);
```

Comment:
According to the [C++ reference](https://en.cppreference.com/w/cpp/string/basic_string/resize) we are missing some exception handling here.

Reply:
Check, it might be be a similar amount of work (and more consistent with the rest of the C codebase) to just use `flb_malloc` here as well, since the `simdutf` functions used are `noexcept`-labelled.